### ISSUE TITLE
gradient vanishing

### ISSUE THREAD

**Issue by kingo233** [Thanos-DB/FullyConvolutionalTransformer#9]
Hi! I am reproducing your impressive work.But I came ito gradient vanishing in the block_1.

And the dice stay at 87.

for details,plz check my FCT-Pytorch repo,and see the release.

**Comment by Thanos-DB** [comment]
Hello kingo233,

in our paper we use the Tensorflow implementation and not the Pytorch one. As such the latter may need a little work to get to the same performance. Did you implement everything or is something still in your pipeline(for example the data generators)?
I see you have changed a few things compared to our implementation (like introducing weight_decay, scaling the grad.). I would suggest for now to stick as close as you can to our Tensroflow implementation till you get comparable results.
Do you plot your train and validation losses? Do you see any signs of overfitting because you are using a different number of att_heads which would make your model considerably larger. Can you train with `att_heads = [2, 2, 2, 2, 2, 2, 2, 2, 2]` as seen in the Tensorflow implementation (see the jupyter notebook)? Also, from a quick read i see you are using `scheduler.step(train_loss)` which is also a factor leading to overfitting because ideally you want to use the validation_loss instead as suggested here https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html#torch.optim.lr_scheduler.ReduceLROnPlateau

Disclaimer: i have read your repo fairly quickly

**Comment by kingo233** [comment]
Hi! Thank you for your reply. I have removed scaling gradient and weight decay,modified attention heads,moved lr scheduler to the validation step.But the dice stay at 0.67...
![6bcc2ebfc36e54659d3265a29916a7ad](https://user-images.githubusercontent.com/29831327/226273581-dd80d955-1aeb-4f1c-b6b2-4a026a10e850.png)

![5711f6bece96f964276efedc9b39de58](https://user-images.githubusercontent.com/29831327/226271907-136b4848-9b12-4829-9242-46171dbc163f.png)
![9aa9e88bac4dedff05af28499fc7dcb7](https://user-images.githubusercontent.com/29831327/226273619-07eeeccc-85bf-423b-a506-24b64a2a42ab.png)
![dcca56123edb79415f0778469f674237](https://user-images.githubusercontent.com/29831327/226273795-5e0d3fc5-36b0-4308-b710-93c977144191.png)



And above is the absolute gradient of some parts of the network.Obviously this gradient vanishing.And this should have nothing to do with data augmentation
